### PR TITLE
Data group rules

### PR DESCRIPTION
**Pull request by alxdarksage**
Splits element rules into a before and after set.
Add two new actions for before sets: displayIf and displayUnless
Add the ability to specify data groups to test against (rather than the answer to a question);
Add two new set operators, "any" and "all", specifically for testing against data groups
Update validation rules

### CODE REVIEW

File: app/org/sagebionetworks/bridge/dynamodb/DynamoSurveyElement.java [hunk 1]
```diff
@@ -116,7 +125,8 @@ public boolean equals(Object obj) {
                 Objects.equals(order, this.order) &&
                 Objects.equals(surveyCompoundKey, this.surveyCompoundKey) &&
                 Objects.equals(type, other.type) &&
-                Objects.equals(rules, other.rules);
+                Objects.equals(afterRules, other.afterRules) &&
+                Objects.equals(beforeRules, other.beforeRules);
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoSurveyElement.java]: Should we add an equalsVerifier test for this?

File: app/org/sagebionetworks/bridge/dynamodb/DynamoSurveyInfoScreen.java [hunk 2]
```diff
@@ -29,7 +29,8 @@ public DynamoSurveyInfoScreen(SurveyElement entry) {
         setIdentifier(entry.getIdentifier());
         setGuid(entry.getGuid());
         setData(entry.getData());
-        setRules(entry.getRules());
+        setBeforeRules(entry.getBeforeRules());
+        setAfterRules(entry.getAfterRules());
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoSurveyInfoScreen.java:34]: Do we have a unit test to make sure these rules are being properly pulled in from the SurveyElement?

File: app/org/sagebionetworks/bridge/dynamodb/DynamoSurveyQuestion.java [hunk 3]
```diff
@@ -30,11 +30,12 @@ public DynamoSurveyQuestion() {
...
-        setIdentifier( entry.getIdentifier() );
-        setGuid( entry.getGuid() );
-        setData( entry.getData() );
-        setRules( entry.getRules() );
+        setType(entry.getType());
+        setIdentifier(entry.getIdentifier());
+        setGuid(entry.getGuid());
+        setData(entry.getData());
+        setBeforeRules(entry.getBeforeRules());
+        setAfterRules(entry.getAfterRules());
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoSurveyQuestion.java:40]: Similarly, do we have a unit test for this?

File: app/org/sagebionetworks/bridge/models/surveys/SurveyElement.java [hunk 4]
```diff
@@ -17,7 +17,8 @@
     public static final String IMAGE_PROPERTY = "image";
     public static final String PROMPT_DETAIL_PROPERTY = "promptDetail";
     public static final String PROMPT_PROPERTY = "prompt";
-    public static final String RULES_PROPERTY = "rules";
+    public static final String BEFORE_RULES_PROPERTY = "beforeRules";
+    public static final String AFTER_RULES_PROPERTY = "afterRules";
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/models/surveys/SurveyElement.java]: Minor: Fields in interfaces are implicitly public static final.

File: app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java [hunk 5]
```diff
@@ -98,21 +133,26 @@ public boolean equals(Object obj) {
                Objects.equals(operator, other.operator) &&
                Objects.equals(value, other.value) &&
                Objects.equals(endSurvey, other.endSurvey) &&
-               Objects.equals(assignDataGroup, other.assignDataGroup);
+               Objects.equals(assignDataGroup, other.assignDataGroup) &&
+               Objects.equals(dataGroups, other.dataGroups);
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java]: hashCode() and equals() are missing displayIf and displayUnless. This is almost certainly incorrect.

File: app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java [hunk 6]
```diff
@@ -31,22 +33,34 @@
...
-            String assignDataGroup) {
+            String assignDataGroup, Set<String> dataGroups, Boolean displayIf, Boolean displayUnless) {
         this.operator = operator;
         this.value = value;
         this.skipToTarget = skipToTarget;
         this.endSurvey = endSurvey;
         this.assignDataGroup = assignDataGroup;
+        this.dataGroups = dataGroups;
+        this.displayIf = displayIf;
+        this.displayUnless = displayUnless;
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java:63]: Do we have JSON serialization tests for these?

File: app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java [hunk 7]
```diff
@@ -98,21 +133,26 @@ public boolean equals(Object obj) {
...
+               Objects.equals(assignDataGroup, other.assignDataGroup) &&
+               Objects.equals(dataGroups, other.dataGroups);
     }
 
     @Override
     public String toString() {
         return "SurveyRule [operator=" + operator + ", value=" + value + ", skipToTarget=" + skipToTarget
-                + ", endSurvey=" + endSurvey + ", assignDataGroup=" + assignDataGroup + "]";
+                + ", endSurvey=" + endSurvey + ", assignDataGroup=" + assignDataGroup + ", dataGroups=" + 
+                dataGroups + "]";
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java]: Minor: displayIf and displayUnless? Also, what happens if any of these are null?

File: app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java [hunk 8]
```diff
@@ -133,12 +173,29 @@ public Builder withEndSurvey(Boolean endSurvey) {
             }
             return this;
         }
+        public Builder withDisplayIf(Boolean displayIf) {
+            if (Boolean.TRUE.equals(displayIf)) {
+                this.displayIf = displayIf;
+            }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java]: Why do we need to check that this is true? Are we concerned that people might accidentally call withDisplayIf(true), and then later call withDisplayIf(false)? If so, why is leaving it to true more intuitive than leaving it to false?

Alternatively, if the goal is to default to false, it would be better for the lower case b boolean (since we'd only have two states anyway) in both the class and the builder (explicitly initialize it to false for documentation/clarity). This way, it'll be consistent with all the other builders, be easier to reason about, and easier to test.

If we _do_ keep this logic, we'll want to add unit tests to make sure that, once set, we can't clear it with withDisplayIf(false) or withDisplayIf(null).

(Here and in withDisplayUnless().)

File: app/org/sagebionetworks/bridge/validators/SurveySaveValidator.java [hunk 9]
```diff
@@ -224,18 +263,30 @@ private void validateOneRuleSet(Errors errors, SurveyRule rule, Set<String> alre
...
-            errors.rejectValue(fieldPath, "must have one and only one action: skipTo, endSurvey, or assignDataGroup");
+        if (rule.getDisplayIf() != null) {
+            actionCount++;
         }
-        if (rule.getAssignDataGroup() != null && !dataGroups.contains(rule.getAssignDataGroup())) {
-            errors.rejectValue(fieldPath, "has a data group '" + rule.getAssignDataGroup()
-                    + "' that is not a valid data group: " + BridgeUtils.COMMA_SPACE_JOINER.join(dataGroups));            
+        if (rule.getDisplayUnless() != null) {
+            actionCount++;
         }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/validators/SurveySaveValidator.java:275]: It doesn't appear that we have any tests that verify that you can't have displayIf and displayUnless at the same time. We also want to avoid having 2^5 tests testing every possible combination, as that would not be maintainable.

My recommendation is to have N tests, one for each action, so that we can verify that they're each being counted by hasOneAction(), as well as a test with no actions.

File: test/org/sagebionetworks/bridge/validators/SurveySaveValidatorTest.java [hunk 10]
```diff
@@ -881,11 +881,11 @@ public void operatorsInInfoScreenRulesAreValidated() {
         info.setPromptDetail("prompt detail");
         info.setIdentifier("identifier");
         info.setGuid("guid");
-        info.setRules(Lists.newArrayList(
+        info.setAfterRules(Lists.newArrayList(
                 new SurveyRule.Builder().withValue("foo").withOperator(Operator.EQ).withEndSurvey(true).build()));
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/validators/SurveySaveValidatorTest.java]: Should we also add beforeRules to verify that both are being tested? (Whether this logically goes here or in beforeRulesAreValidated() is up to you.)

File: app/org/sagebionetworks/bridge/validators/SurveySaveValidator.java [hunk 11]
```diff
@@ -143,77 +144,115 @@ private void validateRules(Errors errors, List<SurveyElement> elements) {
...
+                // Only validate fields that are relevant for the operator, ignore the other.
+                if (SurveyRule.SET_OPERATORS.contains(rule.getOperator())) {
+                    // tests against data groups
+                    if (isEmpty(rule.getDataGroups())) {
+                        errors.rejectValue(fieldPath, "should define one or more data groups");
+                    } else if (!dataGroups.containsAll(rule.getDataGroups())) {
+                        errors.rejectValue(fieldPath,
+                            "contains dataGroups '" + COMMA_SPACE_JOINER.join(rule.getDataGroups())
+                                    + "' that are not valid data groups: " + COMMA_SPACE_JOINER.join(dataGroups));
+                    }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/validators/SurveySaveValidator.java]: Unit tests for these?
* empty data groups for ANY
* invalid data groups for ANY
* at least one invalid rule with ALL to verify that that's being tested

File: app/org/sagebionetworks/bridge/validators/SurveySaveValidator.java [hunk 12]
```diff
@@ -143,77 +144,115 @@ private void validateRules(Errors errors, List<SurveyElement> elements) {
...
+                } else if (valueMissingForOperator(rule)) {
+                    errors.rejectValue(fieldPath, "is required");
+                }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/validators/SurveySaveValidator.java]: Unit test for missing value?

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]

File: app/org/sagebionetworks/bridge/dynamodb/DynamoSurveyElement.java (hunk 1 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/dynamodb/DynamoSurveyElement.java]: I'm going to remove the hashCode/equals methods from the SurveyElement class hierarchy. It's broken, and isn't used, even for tests. Creating hashCode/equals methods for SurveyElement which has to be mutable for Dynamo, and subclassable in a way that changes equality, and includes JsonNode (which has abstract methods that need prefab values)... it's just a quagmire and not worth it. So, removed.

File: app/org/sagebionetworks/bridge/dynamodb/DynamoSurveyInfoScreen.java (hunk 2 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/dynamodb/DynamoSurveyInfoScreen.java:34]: Added tests of the copy constructors for SurveyInfoScreen and SurveyQuestion

File: app/org/sagebionetworks/bridge/dynamodb/DynamoSurveyQuestion.java (hunk 3 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/dynamodb/DynamoSurveyQuestion.java:40]: Added

File: app/org/sagebionetworks/bridge/models/surveys/SurveyElement.java (hunk 4 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/models/surveys/SurveyElement.java]: Removed unnecessary keywords

File: app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java (hunk 6 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java:63]: Added

File: app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java (hunk 5 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java]: Added (these are used for tests and SurveyRuleTest was failing... I'm rerunning all the tests before committing next round of changes).

File: app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java (hunk 7 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java]: Added (when they are null they just display as "someField=null" which is fine for toString).

File: app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java (hunk 8 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java]: If the user sends a property set to false (which they could do), we set the object field to null so that it is not serialized and sent back in future JSON. Only one boolean property can be true at a time, and this ensures only the true property will be in JSON representations. I've added a test to SurveyRuleTest to verify this.

File: app/org/sagebionetworks/bridge/validators/SurveySaveValidator.java (hunk 11 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/validators/SurveySaveValidator.java]: Added

File: app/org/sagebionetworks/bridge/validators/SurveySaveValidator.java (hunk 12 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/validators/SurveySaveValidator.java]: Added, and tests for null value operators that don't need a value (always, declined)

File: app/org/sagebionetworks/bridge/validators/SurveySaveValidator.java (hunk 9 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/validators/SurveySaveValidator.java:275]: I added a test specifically for each action, no actions, and two actions. This forced some refactoring of the test class to reduce the duplication of creating a survey with one element and one rule in it to validate these SurveyRule validations.

File: test/org/sagebionetworks/bridge/validators/SurveySaveValidatorTest.java (hunk 10 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/validators/SurveySaveValidatorTest.java]: I've verified that in the three ways we validate afterRules, we validate beforeRules, I have not validated the every single validation works through both code paths because I have verified that the same code is executed on both (as both are a list of rules).

File: app/org/sagebionetworks/bridge/validators/SurveySaveValidator.java (hunk 9 above)
Comment (DwayneJengSage) [reply app/org/sagebionetworks/bridge/validators/SurveySaveValidator.java:275]: My concern is that we don't know that this code is correctly counting displayIf and displayUnless unless we test it.

File: test/org/sagebionetworks/bridge/validators/SurveySaveValidatorTest.java (hunk 10 above)
Comment (DwayneJengSage) [reply test/org/sagebionetworks/bridge/validators/SurveySaveValidatorTest.java]: I see the test for validateOneRuleSet() and validateSkipToTargetExists(), but not validateOnRuleSetInInfoScreen().

File: app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java (hunk 8 above)
Comment (DwayneJengSage) [reply app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java]: As written, once this is set to true, it can never be cleared. If we want false to be null, we need an else block to set it to null.

File: app/org/sagebionetworks/bridge/validators/SurveySaveValidator.java (hunk 9 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/validators/SurveySaveValidator.java:275]: I've added one test for each action to verify the action makes a valid rule... I can alter the validator to return the count from a protected method, then one test could verify that all actions are being counted.

File: test/org/sagebionetworks/bridge/validators/SurveySaveValidatorTest.java (hunk 10 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/validators/SurveySaveValidatorTest.java]: Okay, I've added before/after rules to the info screen and validated there are errors for both.

File: app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java (hunk 8 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java]: I haven't really ever called this method more than once because it's in the builder... but I can add that if you think it's a problem down the road.

File: test/org/sagebionetworks/bridge/dynamodb/DynamoSurveyQuestionTest.java [hunk 13]
```diff
@@ -1,17 +1,52 @@
...
+        question.setUiHint(UIHint.COMBOBOX);
+        question.setConstraints(new IntegerConstraints());
+        question.setSurveyCompoundKey("surveyCompoundKey");
+        question.setGuid("guid");
+        question.setIdentifier("identifier");
+        question.setType("SurveyQuestion");
+        SurveyRule beforeRule = new SurveyRule.Builder().withDisplayUnless(true).withDataGroups(Sets.newHashSet("foo")).build();
+        SurveyRule afterRule = new SurveyRule.Builder().withOperator(Operator.ALWAYS).withEndSurvey(true).build();
+        question.setBeforeRules(Lists.newArrayList(beforeRule));
+        question.setAfterRules(Lists.newArrayList(afterRule));
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/dynamodb/DynamoSurveyQuestionTest.java:36]: Before and after rules aren't checked in the copy.

File: app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java (hunk 8 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java]: Updated these to explicitly set to null if not true

File: test/org/sagebionetworks/bridge/models/surveys/SurveyRuleTest.java [hunk 14]
```diff
@@ -94,4 +95,58 @@ public void canSerializeAssignDataGroupRule() throws Exception {
...
+    
+    @Test
+    public void canSerializeDisplayIf() throws Exception {
+        SurveyRule displayIf = new SurveyRule.Builder().withOperator(Operator.ANY)
+                .withDataGroups(Sets.newHashSet("foo")).withDisplayIf(Boolean.TRUE).build();
+        
+        JsonNode node = BridgeObjectMapper.get().valueToTree(displayIf);
+        assertEquals("any", node.get("operator").asText());
+        assertEquals("foo", node.get("dataGroups").get(0).asText());
+        assertTrue(node.get("displayIf").asBoolean());
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/models/surveys/SurveyRuleTest.java]: asBoolean() will attempt to parse the JSON value into a boolean. This means we don't know if the serializer is serializing the value as, for example, the int 1 or the string "true". Recommend using booleanValue() instead. (Likewise textValue() instead of asText(), although this matters less often.)

Similarly for canSerializeDisplayUnless().

File: app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java [hunk 15]
```diff
@@ -31,22 +33,34 @@
...
     
     private SurveyRule(Operator operator, Object value, String skipToTarget, Boolean endSurvey,
-            String assignDataGroup) {
+            String assignDataGroup, Set<String> dataGroups, Boolean displayIf, Boolean displayUnless) {
         this.operator = operator;
         this.value = value;
         this.skipToTarget = skipToTarget;
         this.endSurvey = endSurvey;
         this.assignDataGroup = assignDataGroup;
+        this.dataGroups = dataGroups;
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java:61]: I don't see a serialization test for dataGroups.

File: test/org/sagebionetworks/bridge/validators/SurveySaveValidatorTest.java [hunk 16]
```diff
@@ -785,89 +786,55 @@ public void willValidateEarliestDateTimeIsNotAfterLatestDateTime() {
...
     @Test
     public void endSurveyRuleValidInConstraints() {
-        survey = new TestSurvey(SurveySaveValidatorTest.class, false);
-
-        // This rule is a valid "end the survey" rule, and passes validation.
-        SurveyQuestion question = ((TestSurvey) survey).getIntegerQuestion();
         SurveyRule rule = new SurveyRule.Builder().withOperator(SurveyRule.Operator.EQ).withValue(1)
                 .withEndSurvey(Boolean.TRUE).build();
-        question.getConstraints().setRules(Lists.newArrayList(rule));
+        updateSurveyWithAfterRulesInOneQuestion(rule);
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/validators/SurveySaveValidatorTest.java]: This test is for constraints, but got moved to afterRules.

File: test/org/sagebionetworks/bridge/validators/SurveySaveValidatorTest.java [hunk 17]
```diff
@@ -785,89 +786,55 @@ public void willValidateEarliestDateTimeIsNotAfterLatestDateTime() {
...
-        question.getConstraints().setRules(Lists.newArrayList(rule));
+        updateSurveyWithRulesInConstraints(rule);
 
-        assertValidatorMessage(validator, survey, "elements[4].constraints.rules[0]",
+        assertValidatorMessage(validator, survey, "elements[0].constraints.rules[0]",
                 "has a skipTo identifier that doesn't exist: this_does_not_exist");
     }
 
     @Test
-    public void backreferenceSkipToTargetInvalidInElement() throws Exception {
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/validators/SurveySaveValidatorTest.java]: Why is this test deleted?

File: test/org/sagebionetworks/bridge/validators/SurveySaveValidatorTest.java [hunk 18]
```diff
@@ -881,41 +848,218 @@ public void operatorsInInfoScreenRulesAreValidated() {
...
+                "contains dataGroups 'notInStudy' that are not valid data groups: baz, foo");
+    }
+    
+    @Test
+    public void validatesMixedValidityDataGroupsForAllOperator() {
+        TreeSet<String> dataGroups = new TreeSet();
+        dataGroups.add("foo");
+        dataGroups.add("notInStudy");
+        
+        SurveyRule rule = new SurveyRule.Builder().withOperator(SurveyRule.Operator.ANY)
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/validators/SurveySaveValidatorTest.java]: This test is for ALL, but the code is for ANY.

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]

**Review by DwayneJengSage — APPROVED** [review]
